controllers/managedocs_controller.go

```diff
@@ -677,46 +690,76 @@ func (r *ManagedOCSReconciler) getDesiredConvergedStorageCluster() (*ocsv1.Stora
 
 	// Get the storage device set count of the current storage cluster
 	currDeviceSetCount := 0
-	for index := range r.storageCluster.Spec.StorageDeviceSets {
-		item := &r.storageCluster.Spec.StorageDeviceSets[index]
-		if item.Name == deviceSetName {
-			currDeviceSetCount = item.Count
-			break
-		}
+	if desiredStorageDeviceSet := getStorageDeviceSet(r.storageCluster.Spec.StorageDeviceSets, deviceSetName); desiredStorageDeviceSet != nil {
+		currDeviceSetCount = desiredStorageDeviceSet.Count
 	}
 
+	// Get the desired storage device set from storage cluster template
 	sc := templates.StorageClusterTemplate.DeepCopy()
-
 	var ds *ocsv1.StorageDeviceSet = nil
-	for index := range sc.Spec.StorageDeviceSets {
-		item := &sc.Spec.StorageDeviceSets[index]
-		if item.Name == deviceSetName {
-			ds = item
-			break
-		}
-	}
-	if ds == nil {
-		return nil, fmt.Errorf("could not find default device set on stroage cluster")
+	if desiredStorageDeviceSet := getStorageDeviceSet(sc.Spec.StorageDeviceSets, deviceSetName); desiredStorageDeviceSet != nil {
+		ds = desiredStorageDeviceSet
 	}
 
 	// Prevent downscaling by comparing count from secret and count from storage cluster
 	r.Log.Info("Setting storage device set count", "Current", currDeviceSetCount, "New", desiredDeviceSetCount)
-	if currDeviceSetCount <= desiredDeviceSetCount {
-		ds.Count = desiredDeviceSetCount
-	} else {
-		r.Log.V(-1).Info("Requested storage device set count will result in downscaling, which is not supported. Skipping")
-		ds.Count = currDeviceSetCount
-	}
+	r.updateDeviceSetCount(ds, desiredDeviceSetCount, currDeviceSetCount)
+
 	// Check and enable MCG in Storage Cluster spec
-	mcgEnable, err := strconv.ParseBool(enableMCGAsString)
+	if err := r.ensureMCGDeployment(sc, enableMCGAsString); err != nil {
+		return nil, err
+	}
+
+	return sc, nil
+}
+
+func (r *ManagedOCSReconciler) getDesiredProviderStorageCluster() (*ocsv1.StorageCluster, error) {
+	_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, r.onboardingValidationKeySecret, func() error {
+		if err := r.own(r.onboardingValidationKeySecret); err != nil {
+			return err
+		}
+		r.onboardingValidationKeySecret.Data = map[string][]byte{
+			"key": []byte(r.addonParams[onboardingValidationKey]),
+		}
+		return nil
+	})
```

Comment:
Why do we have a `CreateOrUpdate` here? by design, each reconcile function should only reconcile a single output resource. If we need an `reconcileOnboardinValidationSecret` function  then lets add it